test/e2e/apps/deployment.go

```diff
@@ -799,6 +802,132 @@ func testDeploymentsControllerRef(f *framework.Framework) {
+	err = framework.VerifyPodsRunning(c, ns, NginxImageName, false, *(deployment.Spec.Replicas))
+	Expect(err).NotTo(HaveOccurred(), "error in waiting for pods to come up: %v", err)
+
+	framework.Logf("Waiting for deployment %q to complete", deployment.Name)
+	Expect(framework.WaitForDeploymentComplete(c, deployment)).NotTo(HaveOccurred())
+
+	firstRS, err := deploymentutil.GetNewReplicaSet(deployment, c.ExtensionsV1beta1())
+	Expect(err).NotTo(HaveOccurred())
+
+	// Update the deployment with a non-existent image so that the new replica set
+	// will be blocked to simulate a partial rollout.
+	framework.Logf("Updating deployment %q with a non-existent image", deploymentName)
+	deployment, err = framework.UpdateDeploymentWithRetries(c, ns, d.Name, func(update *extensions.Deployment) {
+		update.Spec.Template.Spec.Containers[0].Image = "nginx:404"
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	framework.Logf("Waiting for observed generation %d", deployment.Generation)
+	Expect(framework.WaitForObservedDeployment(c, ns, deploymentName, deployment.Generation)).NotTo(HaveOccurred())
+
+	framework.Logf("Verifying that deployment %q has minimum required number of available replicas", deploymentName)
+	minAvailableReplicas := deploymentutil.MinAvailable(deployment)
+	if deployment.Status.AvailableReplicas < minAvailableReplicas {
+		Expect(fmt.Errorf("observed %d available replicas, less than min required %d", deployment.Status.AvailableReplicas, minAvailableReplicas)).NotTo(HaveOccurred())
+	}
+
+	framework.Logf("Waiting for the replicasets of deployment %q to have desired number of replicas", deploymentName)
+	firstRS, err = c.ExtensionsV1beta1().ReplicaSets(ns).Get(firstRS.Name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	secondRS, err := deploymentutil.GetNewReplicaSet(deployment, c.ExtensionsV1beta1())
+	Expect(err).NotTo(HaveOccurred())
+
+	err = framework.WaitForReplicaSetDesiredReplicas(c.ExtensionsV1beta1(), firstRS)
+	Expect(err).NotTo(HaveOccurred())
+	err = framework.WaitForReplicaSetDesiredReplicas(c.ExtensionsV1beta1(), secondRS)
+	Expect(err).NotTo(HaveOccurred())
+
+	framework.Logf("Verifying that first rollout's replicaset has minimum required number of available replicas")
+	if firstRS.Status.AvailableReplicas < minAvailableReplicas {
+		Expect(fmt.Errorf("observed %d available replicas, less than min required %d", firstRS.Status.AvailableReplicas, minAvailableReplicas)).NotTo(HaveOccurred())
+	}
+
+	maxUnavailable, err := intstr.GetValueFromIntOrPercent(deployment.Spec.Strategy.RollingUpdate.MaxUnavailable, int(*(deployment.Spec.Replicas)), false)
+	Expect(err).NotTo(HaveOccurred())
+
+	// First rollout's replicaset should have .spec.replicas = 10 - 2 = 8 replicas.
+	newSpecReplicas := replicas - int32(maxUnavailable)
+	framework.Logf("Verifying that first rollout's replicaset has .spec.replicas = %d", newSpecReplicas)
+	Expect(*firstRS.Spec.Replicas).Should(Equal(newSpecReplicas))
```

Comment:
This is a state we expect to reach when the Deployment cannot progress further, but we cannot expect this to happen right after we upgrade the Deployment. We'll need to wait. 

Therefore, please change this into a wait poll and move this step up to right after 2nd rollout:
1. Update deployment to the bad (2nd) rollout 
2. Wait until the RS1 has 8 available replicas. Once it's done, check that it has 8 replicas and do the desired replica wait too
   - Check available replicas first because `maxUnavailable` affects the total number of **available replicas** of all RSes. Although the number will be the same as replicas, they still carry different meanings and clarifying it can reduce confusion. 
   - As a side note, the desired replicas wait makes sure that the RS controller has created expected number of pods. 
3. Wait until the RS2 has 5 replicas (Deployment's replicas + maxsurge - RS1's replicas). Once it's done, check that it has 0 available replicas and do the desired replica wait. 
   - Calculate RS2's replicas from RS1's replicas because `maxSurge` affects the total number of **replicas** of all RSes. 
4. Do other checks, such as Deployment's min availability, etc. 
5. Finally, scale the Deployment to start the real test.

Reply:
To reiterate, there is a need to check number of available replicas for both RSs to verify the numbers do not conflict with `maxUnavailable` value. Thank you too for emphasizing differences between `maxUnavailable` (affecting total number of _available replicas_) and `maxSurge` (affecting total number of _replicas_).

Reply:
@janetkuo To reiterate, there is a need to check number of available replicas for both RSs to verify the numbers do not conflict with `maxUnavailable` value. Thank you too for emphasizing differences between `maxUnavailable` (affecting total number of _available replicas_) and `maxSurge` (affecting total number of _replicas_).